File: packages/mobile/ios/App/App/OpenChamberVirtualAssetStore.swift
import Foundation

/// In-memory progressive byte queues for virtual image assets.
/// Opaque asset IDs only — never stores host paths, credentials, or tunnel keys.
enum OpenChamberVirtualAssetStore {
    static let scheme = "openchamber-asset"
    static let urlHost = "v"

    /// Incomplete assets expire this long after last activity.
    static let ttlSeconds: TimeInterval = 120
    static let maxConcurrentAssets = 16
    /// Hard ceiling per asset body.
    static let maxAssetBytes = 32 * 1024 * 1024
    /// Pending unread queue size before append waits / rejects (backpressure).
    static let maxQueuedBytes = 4 * 1024 * 1024
    static let appendWaitSeconds: TimeInterval = 15
    static let assetIdPattern = #"^[A-Za-z0-9_-]{8,80}$"#
    /// image/* only — type/subtype form, optional +suffix, no parameters (matches Electron).
    private static let imageMimePattern = #"^image/[a-z0-9][a-z0-9!#$&\-^_.+]*$"#

    private static let lock = NSLock()
    private static var assets: [String: Asset] = [:]
    private static var pruneTimer: Timer?

    private final class Asset {
        let assetId: String
        let mime: String
        let createdAt: Date
        var lastActivity: Date
        var finished = false
        var cancelled = false
        var totalBytes = 0
        var queuedBytes = 0
        var chunks: [Data] = []
        var waiters: [DispatchSemaphore] = []
        /// Tasks currently reading this asset (scheme handlers).
        var readers: [ObjectIdentifier: Reader] = [:]

        init(assetId: String, mime: String) {
            self.assetId = assetId
            self.mime = mime
            let now = Date()
            self.createdAt = now
            self.lastActivity = now
        }
    }

    final class Reader {
        let id = ObjectIdentifier(NSObject())
        var stopped = false
    }

    enum StoreError: LocalizedError {
        case invalidAssetId
        case invalidMime
        case duplicate
        case limitReached
        case notFound
        case cancelled
        case finished
        case tooLarge
        case backpressure
        case invalidChunk

        var errorDescription: String? {
            switch self {
            case .invalidAssetId: return "A valid assetId is required."
            case .invalidMime: return "A valid mime is required."
            case .duplicate: return "assetId already exists."
            case .limitReached: return "Virtual asset concurrency limit reached."
            case .notFound: return "Unknown assetId."
            case .cancelled: return "Asset was cancelled."
            case .finished: return "Asset already finished."
            case .tooLarge: return "Asset exceeds maximum size."
            case .backpressure: return "Asset queue backpressure timeout."
            case .invalidChunk: return "Invalid base64 chunk."
            }
        }
    }

    static func url(for assetId: String) -> String {
        "\(scheme)://\(urlHost)/\(assetId)"
    }

    static func extractAssetId(from url: URL) -> String? {
        guard url.scheme?.lowercased() == scheme else { return nil }
        let host = (url.host ?? "").lowercased()
        guard host == urlHost else { return nil }
        let parts = url.path.split(separator: "/").map(String.init)
        guard let raw = parts.first, !raw.isEmpty else { return nil }
        return raw.removingPercentEncoding ?? raw
    }

    /// Normalize and validate image-only MIME (image/*, ≤128, no CR/LF/NUL, strict token).
    static func normalizeImageMime(_ mime: String) -> String? {
        let normalized = mime.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if normalized.isEmpty || normalized.count > 128 { return nil }
        if normalized.contains("\n") || normalized.contains("\r") || normalized.contains("\0") { return nil }
        guard normalized.hasPrefix("image/") else { return nil }
        guard normalized.range(of: imageMimePattern, options: [.regularExpression, .caseInsensitive]) != nil else {
            return nil
        }
        return normalized
    }

    static func create(assetId: String, mime: String) throws -> String {
        guard assetId.range(of: assetIdPattern, options: .regularExpression) != nil else {
            throw StoreError.invalidAssetId
        }
        guard let normalizedMime = normalizeImageMime(mime) else {
            throw StoreError.invalidMime
        }

        lock.lock()
        defer { lock.unlock() }
        pruneExpiredLocked(now: Date())
        if assets[assetId] != nil { throw StoreError.duplicate }
        if assets.count >= maxConcurrentAssets { throw StoreError.limitReached }
        assets[assetId] = Asset(assetId: assetId, mime: normalizedMime)
        schedulePruneLocked()
        return url(for: assetId)
    }

    static func append(assetId: String, base64 chunk: String) throws {
        let data = try decodeBase64(chunk)
        if data.isEmpty { return }

        let deadline = Date().addingTimeInterval(appendWaitSeconds)
        while true {
            lock.lock()
            pruneExpiredLocked(now: Date())
            guard let asset = assets[assetId] else {
                lock.unlock()
                throw StoreError.notFound
            }
            if asset.cancelled {
                lock.unlock()
                throw StoreError.cancelled
            }
            if asset.finished {
                lock.unlock()
                throw StoreError.finished
            }
            if asset.totalBytes + data.count > maxAssetBytes {
                lock.unlock()
                throw StoreError.tooLarge
            }
            if asset.queuedBytes + data.count <= maxQueuedBytes {
                asset.chunks.append(data)
                asset.queuedBytes += data.count
                asset.totalBytes += data.count
                asset.lastActivity = Date()
                wakeWaitersLocked(asset)
                lock.unlock()
                return
            }
            // Backpressure: wait for readers to drain.
            let waiter = DispatchSemaphore(value: 0)
            asset.waiters.append(waiter)
            lock.unlock()
            let remaining = deadline.timeIntervalSinceNow
            if remaining <= 0 {
                removeWaiter(assetId: assetId, waiter: waiter)
                throw StoreError.backpressure
            }
            let timeout = DispatchTime.now() + remaining
            if waiter.wait(timeout: timeout) == .timedOut {
                removeWaiter(assetId: assetId, waiter: waiter)
                throw StoreError.backpressure
            }
        }
    }

    static func finish(assetId: String) throws {
        lock.lock()
        defer { lock.unlock() }
        guard let asset = assets[assetId] else { throw StoreError.notFound }
        if asset.cancelled { throw StoreError.cancelled }
        asset.finished = true
        asset.lastActivity = Date()
        wakeWaitersLocked(asset)
    }

    static func cancel(assetId: String) throws {
        lock.lock()
        defer { lock.unlock() }
        guard let asset = assets[assetId] else { return }
        asset.cancelled = true
        asset.finished = true
        asset.chunks.removeAll(keepingCapacity: false)
        asset.queuedBytes = 0
        for reader in asset.readers.values {
            reader.stopped = true
        }
        wakeWaitersLocked(asset)
        // Drop when no active readers remain.
        if asset.readers.isEmpty {
            assets.removeValue(forKey: assetId)
        }
    }

    // MARK: - Scheme handler integration

    /// One consumer per asset — a second reader is rejected so the shared queue is not sharded.
    static func beginRead(assetId: String) -> (mime: String, reader: Reader)? {
        lock.lock()
        defer { lock.unlock() }
        pruneExpiredLocked(now: Date())
        guard let asset = assets[assetId], !asset.cancelled else { return nil }
        if !asset.readers.isEmpty { return nil }
        let reader = Reader()
        asset.readers[ObjectIdentifier(reader)] = reader
        asset.lastActivity = Date()
        return (asset.mime, reader)
    }

    /// Blocks until a chunk is available, the asset finishes empty, or cancel/stop.
    /// Returns nil when the stream ends (finish with no more data, cancel, or stop).
    static func takeChunk(assetId: String, reader: Reader, waitSeconds: TimeInterval = 30) -> Data? {
        let deadline = Date().addingTimeInterval(waitSeconds)
        while true {
            lock.lock()
            guard let asset = assets[assetId] else {
                lock.unlock()
                return nil
            }
            if reader.stopped || asset.cancelled {
                endReaderLocked(assetId: assetId, reader: reader)
                lock.unlock()
                return nil
            }
            if !asset.chunks.isEmpty {
                let data = asset.chunks.removeFirst()
                asset.queuedBytes = max(0, asset.queuedBytes - data.count)
                asset.lastActivity = Date()
                // Unblock append waiters after drain.
                if !asset.waiters.isEmpty {
                    let waiter = asset.waiters.removeFirst()
                    waiter.signal()
                }
                lock.unlock()
                return data
            }
            if asset.finished {
                endReaderLocked(assetId: assetId, reader: reader)
                lock.unlock()
                return nil
            }
            let waiter = DispatchSemaphore(value: 0)
            asset.waiters.append(waiter)
            lock.unlock()
            let remaining = deadline.timeIntervalSinceNow
            if remaining <= 0 {
                removeWaiter(assetId: assetId, waiter: waiter)
                return nil
            }
            _ = waiter.wait(timeout: .now() + remaining)
        }
    }

    static func endRead(assetId: String, reader: Reader) {
        lock.lock()
        endReaderLocked(assetId: assetId, reader: reader)
        lock.unlock()
    }

    static func stopReader(_ reader: Reader) {
        reader.stopped = true
        lock.lock()
        for asset in assets.values {
            if asset.readers[ObjectIdentifier(reader)] != nil {
                wakeWaitersLocked(asset)
            }
        }
        lock.unlock()
    }

    // MARK: - Internals

    private static func endReaderLocked(assetId: String, reader: Reader) {
        guard let asset = assets[assetId] else { return }
        asset.readers.removeValue(forKey: ObjectIdentifier(reader))
        if (asset.finished || asset.cancelled) && asset.readers.isEmpty {
            assets.removeValue(forKey: assetId)
        }
    }

    private static func wakeWaitersLocked(_ asset: Asset) {
        let pending = asset.waiters
        asset.waiters.removeAll(keepingCapacity: false)
        for waiter in pending {
            waiter.signal()
        }
    }

    private static func removeWaiter(assetId: String, waiter: DispatchSemaphore) {
        lock.lock()
        if let asset = assets[assetId] {
            asset.waiters.removeAll { $0 === waiter }
        }
        lock.unlock()
    }

    private static func pruneExpiredLocked(now: Date) {
        let expired = assets.filter { _, asset in
            now.timeIntervalSince(asset.lastActivity) > ttlSeconds
        }
        for (id, asset) in expired {
            asset.cancelled = true
            asset.finished = true
            asset.chunks.removeAll(keepingCapacity: false)
            asset.queuedBytes = 0
            for reader in asset.readers.values {
                reader.stopped = true
            }
            wakeWaitersLocked(asset)
            if asset.readers.isEmpty {
                assets.removeValue(forKey: id)
            }
        }
    }

    private static func schedulePruneLocked() {
        if pruneTimer != nil { return }
        DispatchQueue.main.async {
            lock.lock()
            if pruneTimer == nil {
                let timer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { _ in
                    lock.lock()
                    pruneExpiredLocked(now: Date())
                    if assets.isEmpty {
                        pruneTimer?.invalidate()
                        pruneTimer = nil
                    }
                    lock.unlock()
                }
                RunLoop.main.add(timer, forMode: .common)
                pruneTimer = timer
            }
            lock.unlock()
        }
    }

    private static func decodeBase64(_ chunk: String) throws -> Data {
        let trimmed = chunk.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return Data() }
        // Accept standard and URL-safe base64, with or without padding.
        var normalized = trimmed
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let pad = normalized.count % 4
        if pad != 0 {
            normalized.append(String(repeating: "=", count: 4 - pad))
        }
        guard let data = Data(base64Encoded: normalized, options: [.ignoreUnknownCharacters]) else {
            throw StoreError.invalidChunk
        }
        return data
    }

    #if DEBUG
    static func debugReset() {
        lock.lock()
        assets.removeAll()
        lock.unlock()
    }

    static func debugCount() -> Int {
        lock.lock()
        defer { lock.unlock() }
        return assets.count
    }
    #endif
}

File: packages/mobile/android/app/src/main/java/com/openchamber/app/OpenChamberVirtualAssetStore.java
package com.openchamber.app;

import android.util.Base64;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;
import java.util.regex.Pattern;

/**
 * Thread-safe progressive byte queues for virtual image assets.
 * Opaque asset IDs only — never stores host paths, credentials, or tunnel keys.
 */
final class OpenChamberVirtualAssetStore {
    static final String SCHEME = "openchamber-asset";
    static final String URL_HOST = "v";

    static final long TTL_MILLIS = 120_000L;
    static final int MAX_CONCURRENT_ASSETS = 16;
    static final int MAX_ASSET_BYTES = 32 * 1024 * 1024;
    static final int MAX_QUEUED_BYTES = 4 * 1024 * 1024;
    static final long APPEND_WAIT_MILLIS = 15_000L;
    static final long READ_WAIT_MILLIS = 30_000L;

    private static final Pattern ASSET_ID = Pattern.compile("^[A-Za-z0-9_-]{8,80}$");
    /** image/* only — type/subtype form, optional +suffix, no parameters (matches Electron). */
    private static final Pattern IMAGE_MIME = Pattern.compile(
        "^image/[a-z0-9][a-z0-9!#$&\\-^_.+]*$",
        Pattern.CASE_INSENSITIVE
    );
    private static final ReentrantLock LOCK = new ReentrantLock();
    private static final Map<String, Asset> ASSETS = new HashMap<>();

    private OpenChamberVirtualAssetStore() {}

    static String urlFor(String assetId) {
        return SCHEME + "://" + URL_HOST + "/" + assetId;
    }

    static String extractAssetId(android.net.Uri uri) {
        if (uri == null) return null;
        String scheme = uri.getScheme();
        if (scheme == null || !SCHEME.equalsIgnoreCase(scheme)) return null;
        String host = uri.getHost();
        if (host == null || !URL_HOST.equalsIgnoreCase(host)) return null;
        String path = uri.getPath();
        if (path == null || path.isEmpty()) return null;
        String trimmed = path.startsWith("/") ? path.substring(1) : path;
        int slash = trimmed.indexOf('/');
        if (slash >= 0) trimmed = trimmed.substring(0, slash);
        if (trimmed.isEmpty()) return null;
        try {
            return java.net.URLDecoder.decode(trimmed, "UTF-8");
        } catch (Exception ignored) {
            return trimmed;
        }
    }

    /**
     * Normalize and validate image-only MIME (image/*, ≤128, no CR/LF/NUL, strict token).
     * @return lowercase normalized type, or null when rejected
     */
    static String normalizeImageMime(String mime) {
        if (mime == null) return null;
        String normalized = mime.trim().toLowerCase(java.util.Locale.ROOT);
        if (normalized.isEmpty() || normalized.length() > 128) return null;
        if (normalized.indexOf('\n') >= 0 || normalized.indexOf('\r') >= 0 || normalized.indexOf('\0') >= 0) {
            return null;
        }
        if (!normalized.startsWith("image/")) return null;
        if (!IMAGE_MIME.matcher(normalized).matches()) return null;
        return normalized;
    }

    static String create(String assetId, String mime) throws Exception {
        if (assetId == null || !ASSET_ID.matcher(assetId).matches()) {
            throw new IllegalArgumentException("A valid assetId is required.");
        }
        String normalizedMime = normalizeImageMime(mime);
        if (normalizedMime == null) {
            throw new IllegalArgumentException("A valid mime is required.");
        }
        LOCK.lock();
        try {
            pruneExpiredLocked(System.currentTimeMillis());
            if (ASSETS.containsKey(assetId)) {
                throw new IllegalStateException("assetId already exists.");
            }
            if (ASSETS.size() >= MAX_CONCURRENT_ASSETS) {
                throw new IllegalStateException("Virtual asset concurrency limit reached.");
            }
            ASSETS.put(assetId, new Asset(assetId, normalizedMime));
            return urlFor(assetId);
        } finally {
            LOCK.unlock();
        }
    }

    static void append(String assetId, String base64Chunk) throws Exception {
        byte[] data = decodeBase64(base64Chunk);
        if (data.length == 0) return;

        long deadline = System.currentTimeMillis() + APPEND_WAIT_MILLIS;
        LOCK.lock();
        try {
            while (true) {
                pruneExpiredLocked(System.currentTimeMillis());
                Asset asset = ASSETS.get(assetId);
                if (asset == null) throw new IllegalStateException("Unknown assetId.");
                if (asset.cancelled) throw new IllegalStateException("Asset was cancelled.");
                if (asset.finished) throw new IllegalStateException("Asset already finished.");
                if ((long) asset.totalBytes + data.length > MAX_ASSET_BYTES) {
                    throw new IllegalStateException("Asset exceeds maximum size.");
                }
                if (asset.queuedBytes + data.length <= MAX_QUEUED_BYTES) {
                    asset.chunks.addLast(data);
                    asset.queuedBytes += data.length;
                    asset.totalBytes += data.length;
                    asset.lastActivity = System.currentTimeMillis();
                    asset.condition.signalAll();
                    return;
                }
                long remaining = deadline - System.currentTimeMillis();
                if (remaining <= 0) {
                    throw new IllegalStateException("Asset queue backpressure timeout.");
                }
                asset.condition.await(remaining, TimeUnit.MILLISECONDS);
            }
        } finally {
            LOCK.unlock();
        }
    }

    static void finish(String assetId) throws Exception {
        LOCK.lock();
        try {
            Asset asset = ASSETS.get(assetId);
            if (asset == null) throw new IllegalStateException("Unknown assetId.");
            if (asset.cancelled) throw new IllegalStateException("Asset was cancelled.");
            asset.finished = true;
            asset.lastActivity = System.currentTimeMillis();
            asset.condition.signalAll();
        } finally {
            LOCK.unlock();
        }
    }

    static void cancel(String assetId) {
        LOCK.lock();
        try {
            Asset asset = ASSETS.get(assetId);
            if (asset == null) return;
            asset.cancelled = true;
            asset.finished = true;
            asset.chunks.clear();
            asset.queuedBytes = 0;
            asset.condition.signalAll();
            if (asset.activeReaders == 0) {
                ASSETS.remove(assetId);
            }
        } finally {
            LOCK.unlock();
        }
    }

    /**
     * Blocking InputStream that yields progressive bytes from the asset queue.
     * One consumer per asset — a second open is rejected so the shared queue is not sharded.
     */
    static InputStream openStream(String assetId) throws IOException {
        LOCK.lock();
        try {
            pruneExpiredLocked(System.currentTimeMillis());
            Asset asset = ASSETS.get(assetId);
            if (asset == null || asset.cancelled) {
                throw new IOException("Unknown or cancelled assetId.");
            }
            if (asset.activeReaders > 0) {
                throw new IOException("Asset already has an active reader.");
            }
            asset.activeReaders += 1;
            asset.lastActivity = System.currentTimeMillis();
            return new AssetInputStream(assetId, asset);
        } finally {
            LOCK.unlock();
        }
    }

    static String mimeOf(String assetId) {
        LOCK.lock();
        try {
            Asset asset = ASSETS.get(assetId);
            return asset != null ? asset.mime : null;
        } finally {
            LOCK.unlock();
        }
    }

    private static void releaseReader(String assetId) {
        LOCK.lock();
        try {
            Asset asset = ASSETS.get(assetId);
            if (asset == null) return;
            asset.activeReaders = Math.max(0, asset.activeReaders - 1);
            if ((asset.finished || asset.cancelled) && asset.activeReaders == 0) {
                ASSETS.remove(assetId);
            }
        } finally {
            LOCK.unlock();
        }
    }

    private static void pruneExpiredLocked(long now) {
        Iterator<Map.Entry<String, Asset>> it = ASSETS.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry<String, Asset> entry = it.next();
            Asset asset = entry.getValue();
            if (now - asset.lastActivity <= TTL_MILLIS) continue;
            asset.cancelled = true;
            asset.finished = true;
            asset.chunks.clear();
            asset.queuedBytes = 0;
            asset.condition.signalAll();
            if (asset.activeReaders == 0) {
                it.remove();
            }
        }
    }

    private static byte[] decodeBase64(String chunk) {
        if (chunk == null) throw new IllegalArgumentException("Invalid base64 chunk.");
        String trimmed = chunk.trim();
        if (trimmed.isEmpty()) return new byte[0];
        String normalized = trimmed.replace('-', '+').replace('_', '/');
        int pad = normalized.length() % 4;
        if (pad != 0) {
            StringBuilder builder = new StringBuilder(normalized);
            for (int i = 0; i < 4 - pad; i++) builder.append('=');
            normalized = builder.toString();
        }
        try {
            return Base64.decode(normalized, Base64.DEFAULT);
        } catch (IllegalArgumentException error) {
            throw new IllegalArgumentException("Invalid base64 chunk.", error);
        }
    }

    private static final class Asset {
        final String assetId;
        final String mime;
        final ArrayDeque<byte[]> chunks = new ArrayDeque<>();
        final Condition condition = LOCK.newCondition();
        long lastActivity = System.currentTimeMillis();
        boolean finished = false;
        boolean cancelled = false;
        int totalBytes = 0;
        int queuedBytes = 0;
        int activeReaders = 0;

        Asset(String assetId, String mime) {
            this.assetId = assetId;
            this.mime = mime;
        }
    }

    /**
     * Thread-safe blocking InputStream. WebView may read this off the UI thread;
     * {@link #read} waits on the asset condition when the queue is empty.
     */
    private static final class AssetInputStream extends InputStream {
        private final String assetId;
        private final Asset asset;
        private byte[] current;
        private int offset;
        private boolean closed;

        AssetInputStream(String assetId, Asset asset) {
            this.assetId = assetId;
            this.asset = asset;
        }

        @Override
        public int read() throws IOException {
            byte[] one = new byte[1];
            int n = read(one, 0, 1);
            return n <= 0 ? -1 : (one[0] & 0xff);
        }

        @Override
        public int read(byte[] buffer, int off, int len) throws IOException {
            if (closed) throw new IOException("Stream closed.");
            if (buffer == null) throw new NullPointerException();
            if (off < 0 || len < 0 || off + len > buffer.length) {
                throw new IndexOutOfBoundsException();
            }
            if (len == 0) return 0;

            LOCK.lock();
            try {
                long deadline = System.currentTimeMillis() + READ_WAIT_MILLIS;
                while (true) {
                    if (closed || asset.cancelled) {
                        return -1;
                    }
                    if (current != null && offset < current.length) {
                        int available = current.length - offset;
                        int copy = Math.min(available, len);
                        System.arraycopy(current, offset, buffer, off, copy);
                        offset += copy;
                        if (offset >= current.length) {
                            current = null;
                            offset = 0;
                        }
                        asset.lastActivity = System.currentTimeMillis();
                        // Drain reduces queued bytes for backpressure.
                        asset.condition.signalAll();
                        return copy;
                    }
                    if (!asset.chunks.isEmpty()) {
                        current = asset.chunks.removeFirst();
                        offset = 0;
                        asset.queuedBytes = Math.max(0, asset.queuedBytes - current.length);
                        asset.lastActivity = System.currentTimeMillis();
                        asset.condition.signalAll();
                        continue;
                    }
                    if (asset.finished) {
                        return -1;
                    }
                    long remaining = deadline - System.currentTimeMillis();
                    if (remaining <= 0) {
                        // Idle timeout: treat as end so WebView does not hang forever.
                        return -1;
                    }
                    try {
                        asset.condition.await(remaining, TimeUnit.MILLISECONDS);
                    } catch (InterruptedException interrupted) {
                        Thread.currentThread().interrupt();
                        throw new IOException("Interrupted while reading virtual asset.", interrupted);
                    }
                }
            } finally {
                LOCK.unlock();
            }
        }

        @Override
        public void close() {
            if (closed) return;
            closed = true;
            LOCK.lock();
            try {
                current = null;
                offset = 0;
                asset.condition.signalAll();
            } finally {
                LOCK.unlock();
            }
            releaseReader(assetId);
        }
    }
}
